### PR TITLE
feat: glibc extension

### PR DESCRIPTION
**Pull request by jfroy**
This PR adds a glibc extension. The intention is to replace the nvidia extensions entirely and only provide the glibc components required by the nvidia gpu-operator and its components (e.g. the nvidia container toolkit).

The extension is pretty much copied from the package that is in the nvidia extensions, with one major modification: a symbolic link to ldconfig is installed at `/sbin/ldconfig`. This change allows the nvidia gpu-operator to work without modification*. This change does require a patch to the extension validation logic, which is provided in separate PRs.

* A patch to the nvidia container toolkit is required to replace the shell script wrappers with a Go wrapper. See https://github.com/NVIDIA/nvidia-container-toolkit/pull/700.

### CODE REVIEW

**Comment by frezbo** [comment]
> This PR adds a glibc extension. The intention is to replace the nvidia extensions entirely and only provide the glibc components required by the nvidia gpu-operator and its components (e.g. the nvidia container toolkit).
> 
> The extension is pretty much copied from the package that is in the nvidia extensions, with one major modification: a symbolic link to ldconfig is installed at `/sbin/ldconfig`. This change allows the nvidia gpu-operator to work without modification*. This change does require a patch to the extension validation logic, which is provided in separate PRs.
> 
>     * A patch to the nvidia container toolkit is required to replace the shell script wrappers with a Go wrapper. See [Replace shell wrapper with a Go wrapper NVIDIA/nvidia-container-toolkit#700](https://github.com/NVIDIA/nvidia-container-toolkit/pull/700).

This is really cool and looking forward to actually not having to maintain patches or wrappers. What i still don't understand is about kernel modules, would SideroLabs be still shipping them as extensions (I believe that's the case since only machined in talos can load modules)?

File: README.md [hunk 1]
```diff
@@ -82,6 +82,7 @@ cosign verify --certificate-identity-regexp '@siderolabs\.com$' --certificate-oi
 | Name                            | Image                                                                                                 | Description                 | Version Format  |
 | ------------------------------- | ----------------------------------------------------------------------------------------------------- | --------------------------- | --------------- |
 | [binfmt-misc](misc/binfmt-misc) | [ghcr.io/siderolabs/binfmt-misc](https://github.com/siderolabs/extensions/pkgs/container/binfmt-misc) | Miscellaneous Binary Format | `talos version` |
+| [glibc](misc/glibc)             | [ghcr.io/siderolabs/glibc](https://github.com/siderolabs/extensions/pkgs/container/glibc)             | glibc                       | `talos version` |
```
Comment (frezbo) [inline README.md:85]: actually this can be the glibc version itself, i don't think there's any constraint to any specific talos version

**Comment by jfroy** [comment]
> > This PR adds a glibc extension. The intention is to replace the nvidia extensions entirely and only provide the glibc components required by the nvidia gpu-operator and its components (e.g. the nvidia container toolkit).
> > The extension is pretty much copied from the package that is in the nvidia extensions, with one major modification: a symbolic link to ldconfig is installed at `/sbin/ldconfig`. This change allows the nvidia gpu-operator to work without modification*. This change does require a patch to the extension validation logic, which is provided in separate PRs.
> > ```
> > * A patch to the nvidia container toolkit is required to replace the shell script wrappers with a Go wrapper. See [Replace shell wrapper with a Go wrapper NVIDIA/nvidia-container-toolkit#700](https://github.com/NVIDIA/nvidia-container-toolkit/pull/700).
> > ```
> 
> This is really cool and looking forward to actually not having to maintain patches or wrappers. What i still don't understand is about kernel modules, would SideroLabs be still shipping them as extensions (I believe that's the case since only machined in talos can load modules)?

I am sending more PRs, but this is one of the major changes from an architecture POV: the gpu-operator would be allowed to load and unload kernel modules, which means enabling module unloading in the kernel (see https://github.com/siderolabs/pkgs/pull/1031) and not removing SYS_MODULE from containers (see https://github.com/siderolabs/talos/pull/9339).

Another important note is that only CDI mode (see https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/cdi.html) works on Talos with this patch. The "legacy" runtime hook require more libraries to be present on the system, whereas the CDI hook is a pure Go program that only requires the glibc dynamic loader and `/sbin/ldconfig`.

**Comment by frezbo** [comment]
> Another important note is that only CDI mode (see https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/cdi.html) works on Talos with this patch. The "legacy" runtime hook require more libraries to be present on the system, whereas the CDI hook is a pure Go program that only requires the glibc dynamic loader and /sbin/ldconfig.

We're open to moving to using CDI :+1:

**Comment by jfroy** [comment]
See https://github.com/siderolabs/talos/pull/9339 for main discussion on loading kernel modules.

**Comment by jfroy** [comment]
Patch has been updated to rework the glibc subtree to look like a [merged `/usr`](https://systemd.io/THE_CASE_FOR_THE_USR_MERGE/) root.

File: misc/glibc/manifest.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+version: v1alpha1
+metadata:
+  name: glibc
+  version: "$VERSION"
+  author: Jean-Francois Roy
+  description: |
+    This system extension provides glibc.
+  compatibility:
+    talos:
+      version: ">= v1.5.0"
```
Comment (frezbo) [inline misc/glibc/manifest.yaml:10]: this needs to be 1.9.0 since `/sbin/ldconfig` is allowed from that version

File: misc/glibc/vars.yaml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+VERSION: 2.40
```
Comment (frezbo) [inline misc/glibc/vars.yaml:1]: also copy over the renovate comments

**Comment by smira** [comment]
I'm going to cut v1.9.0-alpha.0 of Talos, bump the extension validator, and then get back to this PR, thank you for your patience!

**Comment by jfroy** [comment]
> I'm going to cut v1.9.0-alpha.0 of Talos, bump the extension validator, and then get back to this PR, thank you for your patience!

No rush, thank you for considering these PRs!

**Comment by frezbo** [comment]
@jfroy is this good to go, I see the referenced PR for nvidia-container-toolkit is still not merged :thinking:

**Comment by jfroy** [comment]
> @jfroy is this good to go, I see the referenced PR for nvidia-container-toolkit is still not merged :thinking: 

Yeah I'm working on that internally. The team is just very busy with other work. We are focusing in particular on CDI and DRA. Sidero could package the toolkit in an extension -- the wrappers are created by an installer component that is invoked by the operator and are not an intrinsic part of the toolkit.

In any case, this patch is needed for the CDI hook to run, so it's good to pick up no matter what.

**Comment by frezbo** [comment]
I'll wait over the weekend and see if our daily runs for nvidia tests works with this change

**Pull request merged**